File: Day12/Hard/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Spreadsheet {
private:
    vector<vector<int>> vec;
public:
    Spreadsheet(int rows) {
        vec.resize(rows, vector<int>(26, 0));
    }

    void setCell(string cell, int value) {
        int row = decode(cell) - 1;
        int col = cell[0] - 'A';
        vec[row][col] = value;
    }

    void resetCell(string cell) {
        setCell(cell, 0);
    }

    int getValue(string formula) {
        int indexOfPlus = formula.find('+');
        string num1 = formula.substr(1, indexOfPlus - 1);
        string num2 = formula.substr(indexOfPlus + 1);
        int ans = 0;
        if (num1[0] >= 'A' && num1[0] <= 'Z') {
            ans += vec[decode(num1) - 1][num1[0] - 'A'];
        } else {
            ans += decode(num1);
        }
        if (num2[0] >= 'A' && num2[0] <= 'Z') {
            ans += vec[decode(num2) - 1][num2[0] - 'A'];
        } else {
            ans += decode(num2);
        }
        return ans;
    }

    int decode(string cell) {
        if (cell[0] >= 'A' && cell[0] <= 'Z') {
            return stoi(cell.substr(1));
        } else {
            return stoi(cell);
        }
    }
};

int main() {
    Spreadsheet spreadsheet(3);
    cout << spreadsheet.getValue("=5+7") << "\n";
    spreadsheet.setCell("A1", 10);
    cout << spreadsheet.getValue("=A1+6") << "\n";
    spreadsheet.setCell("B2", 15);
    cout << spreadsheet.getValue("=A1+B2") << "\n";
    spreadsheet.resetCell("A1");
    cout << spreadsheet.getValue("=A1+B2") << "\n";

    return 0;
}

File: Day12/Hard/Solution.py
class Spreadsheet(object):
    def __init__(self, rows):
        self.vec = [[0] * 26 for _ in range(rows)]

    def setCell(self, cell, value):
        row = self.decode(cell) - 1
        col = ord(cell[0]) - ord('A')
        self.vec[row][col] = value

    def resetCell(self, cell):
        self.setCell(cell, 0)

    def getValue(self, formula):
        indexOfPlus = formula.find('+')
        num1 = formula[1:indexOfPlus]
        num2 = formula[indexOfPlus + 1:]
        ans = 0
        if 'A' <= num1[0] <= 'Z':
            ans += self.vec[self.decode(num1) - 1][ord(num1[0]) - ord('A')]
        else:
            ans += self.decode(num1)
        if 'A' <= num2[0] <= 'Z':
            ans += self.vec[self.decode(num2) - 1][ord(num2[0]) - ord('A')]
        else:
            ans += self.decode(num2)
        return ans

    def decode(self, cell):
        if 'A' <= cell[0] <= 'Z':
            return int(cell[1:])
        else:
            return int(cell)


if __name__ == "__main__":
    spreadsheet = Spreadsheet(3)
    print(spreadsheet.getValue("=5+7"))
    spreadsheet.setCell("A1", 10)
    print(spreadsheet.getValue("=A1+6"))
    spreadsheet.setCell("B2", 15)
    print(spreadsheet.getValue("=A1+B2"))
    spreadsheet.resetCell("A1")
    print(spreadsheet.getValue("=A1+B2"))
